### PR TITLE
Create more informational assert collection log for FunctionalTests o…

### PR DESCRIPTION
**Pull request by lateapexearlyspeed**
…f FileSystemGlobbing.

Hello, this PR is for #44281 .

I verified in my local pc with this helper method in that test project by intentionally changing expected count or updating one item value. It will give more information than before. Also seems some other runtime projects were already using Assert.Collection(). I wrap this method into a 'private static void AssertCollection(IEnumerable<string> expected, IEnumerable<string> actual)'  and use it in specified test file.

Please review and give suggestion if need, so that I can update code correspondingly, thanks very much and hope have a try to contribute .Net !

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=44350) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=44350) before we can accept your contribution.<br/><br/>:x: lateapexearlyspeed [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=44350)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=44350) it.</sub>

File: src/libraries/Microsoft.Extensions.FileSystemGlobbing/tests/FunctionalTests.cs
Comment (danmosemsft) [inline src/libraries/Microsoft.Extensions.FileSystemGlobbing/tests/FunctionalTests.cs:524]: I wonder if it would make sense instead to add something to this so others could use it elsewhere?
https://github.com/dotnet/runtime/blob/c87e75e8ef260ccc6e979f9df1cc012536ee036f/src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs

Eg something like `SetEquals(IEnumerable<T> expected, IEnumerable<T> actual, IComparer<T> comparer)`

I see in there there is already `public static void Equal<T>(HashSet<T> expected, HashSet<T> actual)` in therebut you'd have to make HashSets with the comparer. Plus it doesn't print a useful output (it dumps the whole lot)
AssertExtensions.Equal(new Hashset(expected), new Hashset(actual

Comment (danmosemsft) [reply src/libraries/Microsoft.Extensions.FileSystemGlobbing/tests/FunctionalTests.cs:524]: I don’t think you could implant out with a Hashset though as it doesn’t allow duplicates.

Comment (lateapexearlyspeed) [reply src/libraries/Microsoft.Extensions.FileSystemGlobbing/tests/FunctionalTests.cs]: Hi @danmosemsft , thanks! I fixed comment by creating CollectionEqual() in common AssertExtensions and use it in FunctionalTest.cs.
`public static void CollectionEqual<T>(IEnumerable<T> expected, IEnumerable<T> actual, IEqualityComparer<T> comparer)`
BTW, I have not updated existing HashSetEqual function, please correct me if there is misunderstanding.
`public static void Equal<T>(HashSet<T> expected, HashSet<T> actual)`

Comment (danmosemsft) [reply src/libraries/Microsoft.Extensions.FileSystemGlobbing/tests/FunctionalTests.cs]: I guess what I had in mind was that CollectionEqual would not require sorted inputs, but instead would do something simple like read both into arrays (not HashSet as they could have duplicates), assert array sizes match, then do a simple O(n^2) check that everything in the first list is found in the second. thoughts?

**Comment by danmosemsft** [comment]
I'll let @maryamariyan review this one. :)

File: src/libraries/Microsoft.Extensions.FileSystemGlobbing/tests/FunctionalTests.cs
Comment (lateapexearlyspeed) [reply src/libraries/Microsoft.Extensions.FileSystemGlobbing/tests/FunctionalTests.cs]: @danmosemsft @maryamariyan Got the requirement now and changed code, thanks. 

There are 3 failure cases as follows:
> Xunit.Sdk.XunitException
Expected count: 3
Actual count: 2

> Xunit.Sdk.XunitException
Not found: source4.cs

> Xunit.Sdk.XunitException
Expected more: source2.cs
Actual: totally 2 source2.cs

I used hash bucket (item value -> its count) to manage actual content, so that it should be better than O(n^2) in long list case maybe, and would like to use ValueTuple but sorry I cannot find way how to refer it in this project so I created private struct ItemCount for item information: Original and Remain.

Comment (lateapexearlyspeed) [reply src/libraries/Microsoft.Extensions.FileSystemGlobbing/tests/FunctionalTests.cs]: Hi @maryamariyan , just want to ask if there is more comment so that I can continue updating? I am new at this repo and hope to learn knowledge of dotnet repo by trying to write code inside this repo. Comment welcome, thanks :)

File: src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs
Comment (maryamariyan) [inline src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs:405]: Link to: [design-guidelines/struct](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/struct)

```
❌ DO NOT define mutable value types.
```

perhaps it would be more maintainable if you just have it be a class in this case and then you won't need to reassign the struct here each time the value is modified.

File: src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs
Comment (lateapexearlyspeed) [reply src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs]: Fixed, thanks.

File: src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs
Comment (lateapexearlyspeed) [reply src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs]: Hi @maryamariyan , Fixed, thanks.

**Comment by lateapexearlyspeed** [comment]
Hi @maryamariyan  Seems CI pipeline was blocked by CI issue and was mentioned in #32805 . Is there any way I can re-trigger CI build besides of pushing a new commit, what is action I should do now, thanks.

**Comment by danmosemsft** [comment]
You can close and reopen the PR.

**Comment by lateapexearlyspeed** [comment]
Hi @danmosemsft Thanks for tip :) All checks passed now. Any more comment, can it be merged ?

**Comment by danmosemsft** [comment]
@maryamariyan ?

**Comment by lateapexearlyspeed** [comment]
Hi @maryamariyan Any chance you can have a look at this PR? :)

**Comment by maryamariyan** [comment]
Thanks for the update @lateapexearlyspeed. Looking now

File: src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs
Comment (maryamariyan) [inline src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs:421]: nit: casing on "Not found"

File: src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs
Comment (maryamariyan) [reply src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs:408]: OK it seems like you need this to get the length of expected.

File: src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs
Comment (maryamariyan) [inline src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs:426]: The message could be clearer. I think it's enough to mention that the collections are not equal.

**Review by maryamariyan — APPROVED** [review]
The approach LGTM with some nit comments.
Thanks @lateapexearlyspeed for the contribution. We'll make sure next follow ups are faster :)

File: src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs
Comment (lateapexearlyspeed) [reply src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs]: Fixed, thanks!

File: src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs
Comment (lateapexearlyspeed) [reply src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs]: Fixed by refining exception message, but still keep some non-matched count info because I feel this issue is just to try resolve similar situation. Thanks @maryamariyan !

Excited to see you @maryamariyan have approved it :) But not sure who should be able to merge it as seems I have no write permission to do that.

**Comment by danmosemsft** [comment]
@maryamariyan is this good to merge?

**Comment by maryamariyan** [comment]
It's merged now. Thank you!

File: src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs
Comment (maryamariyan) [inline src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs:426]: Nit: The last ` {currentExpectedItem}` is redundant but that's alright.

**Pull request merged**
